File: 백준/Silver/1991. 트리 순회/트리 순회.cc
#include <iostream>
using namespace std;
const char nl = '\n';
int n;
int lc[28];
int rc[28];
void preorder(int cur){
  cout << (char)(cur + 'A');
  if(lc[cur]) preorder(lc[cur]);
  if(rc[cur]) preorder(rc[cur]);
}
void inorder(int cur){
  if(lc[cur]) inorder(lc[cur]);
  cout << (char)(cur + 'A');
  if(rc[cur]) inorder(rc[cur]);
}
void postorder(int cur){
  if(lc[cur]) postorder(lc[cur]);
  if(rc[cur]) postorder(rc[cur]);
  cout << (char)(cur + 'A'); 
}
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  cin >> n;
  char p, l, r;
  while(n--){
    cin >> p >> l >> r;
    if(l != '.') lc[p-'A'] = l - 'A';
    if(r != '.') rc[p-'A'] = r - 'A';
  }
  preorder(0);
  cout << nl;
  inorder(0);
  cout << nl;
  postorder(0);
}

File: 백준/Silver/1991. 트리 순회/트리 순회.py
#Authored by : chjh2129
import io, os, sys
from collections import defaultdict

# input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline
input = sys.stdin.readline

def preorder(cur: str, G: defaultdict[ str, list[str]] ) -> None:
  if cur == '.':
    return
  print(cur, end='')
  preorder(G[cur][0], G)
  preorder(G[cur][1], G)

def inorder(cur: str, G: defaultdict[ str, list[str] ]) -> None:
  if cur == '.':
    return
  inorder(G[cur][0], G)
  print(cur, end='')
  inorder(G[cur][1], G)

def postorder(cur: str, G: defaultdict[ str, list[str] ]) -> None:
  if cur == '.':
    return
  postorder(G[cur][0], G)
  postorder(G[cur][1], G)
  
  print(cur, end='')

def main() -> None:
  # input
  N = int(input())
  
  # 인접 리스트 선언
  G: defaultdict[str, list[str]] = defaultdict(list)

  for _ in range(N):
    cur, *child = input().rstrip().split()
    G[cur] = child

  # output
  preorder('A', G)
  print()
  inorder('A', G)
  print()
  postorder('A', G)
  


sys.exit(main())